### PR TITLE
Optimizing Repair-ship code. Adding max healing limitation.

### PR DESCRIPTION
**Pull request by Aingar**
Note: healingrates.cfg will move from data/equipment to exe/flhook_plugins folder

### CODE REVIEW

File: Plugins/Public/alley/PlayerRestrictions.cpp [hunk 1]
```diff
@@ -260,16 +268,27 @@ void LoadSettings()
 		ini.close();
 	}
 
-	string scHealingCfgFile = string(szCurDir) + "\\..\\DATA\\EQUIPMENT\\healingrates.cfg";
```
Comment (dsyalex) [inline Plugins/Public/alley/PlayerRestrictions.cpp:263]: have we ensured the live server is ready for this change?

File: Plugins/Public/alley/PlayerRestrictions.cpp [hunk 2]
```diff
@@ -997,62 +1023,48 @@ bool ExecuteCommandString_Callback(CCmds* cmds, const wstring &wscCmd)
...
-			if (TheShipArch->iShipClass == 19)
-			{
-				float curr, max;
-				pub::SpaceObj::GetHealth(iDmgToSpaceID, curr, max);
-				float expecteddmg = (float)1;
-				float projecteddamage = curr - damage;
+	if (iDmgMunitionID != repairMunitionId
+		|| !iDmgTo
+		|| !dmg->is_inflictor_a_player()
+		|| p1 == 65521)
```
Comment (dsyalex) [inline Plugins/Public/alley/PlayerRestrictions.cpp:1033]: what is this unknown variable and this magic number?

File: Plugins/Public/alley/PlayerRestrictions.cpp [hunk 3]
```diff
@@ -997,62 +1023,48 @@ bool ExecuteCommandString_Callback(CCmds* cmds, const wstring &wscCmd)
...
+		return;
 
-				//If a repair gun heals the ship but this doesn't show up, it's because it's hitting the shield.
-				//HkMsgU(L"DEBUG: damage by repair ship, is it healing?");					
-				//PrintUserCmdText(client, L"Projected damage: %f", projecteddamage);
+	if (!iDmgToSpaceID)
```
Comment (dsyalex) [inline Plugins/Public/alley/PlayerRestrictions.cpp:1036]: why is this no longer triggering an immediate return?

File: Plugins/Public/alley/PlayerRestrictions.cpp [hunk 4]
```diff
@@ -997,62 +1023,48 @@ bool ExecuteCommandString_Callback(CCmds* cmds, const wstring &wscCmd)
...
+	{
+		pub::Player::GetShip(iDmgTo, iDmgToSpaceID);
+	}
 
-				if ((projecteddamage <= 1) && (projecteddamage > 0))
-				{
-					//Handle the healing.
-					returncode = SKIPPLUGINS_NOFUNCTIONCALL;
+	float curr, maxHP;
```
Comment (dsyalex) [inline Plugins/Public/alley/PlayerRestrictions.cpp:1041]: let's move this GetHealth stuff below the unconfigured-shipclass return below

File: Plugins/Public/alley/PlayerRestrictions.cpp [hunk 5]
```diff
@@ -997,62 +1023,48 @@ bool ExecuteCommandString_Callback(CCmds* cmds, const wstring &wscCmd)
...
-					{
-						//HkMsgU(L"DEBUG: Health less max");
-						dmg->add_damage_entry(1, testhealth, (DamageEntry::SubObjFate)0);
-						return;
-					}
-				}
-				//else do nothing, means it isn't a healing call.
-			}
-		}
+	float healedHP = (((healing.healingMultiplier / 100) + (curr / maxHP)) * maxHP) + healing.healingStatic;
```
Comment (dsyalex) [inline Plugins/Public/alley/PlayerRestrictions.cpp:1055]: We can eliminate one floating point division by simplifying:
```suggestion
	float healedHP = (((healing.healingMultiplier / 100 * maxHP) + curr)) + healing.healingStatic;
```

File: Plugins/Public/alley/PlayerRestrictions.cpp [hunk 6]
```diff
@@ -997,62 +1023,48 @@ bool ExecuteCommandString_Callback(CCmds* cmds, const wstring &wscCmd)
...
-					// no need to check healingAdditions here as it is set at the same time as healingMultipliers
-					amounttoheal = max / 100 * healingMultipliers[TheShipArchHealed->iShipClass] + healingAdditions[TheShipArchHealed->iShipClass];
+	const auto& healingData = healingMultipliers.find(TheShipArchHealed->iShipClass);
+	if (healingData == healingMultipliers.end())
+		return;
 
-					float testhealth = curr + amounttoheal;
+	const HEALING_DATA& healing = healingData->second;
 
+	if (curr / maxHP >= healing.maxHeal)
```
Comment (dsyalex) [inline Plugins/Public/alley/PlayerRestrictions.cpp:1052]: Can avoid a floating point division here:
```suggestion
	if (curr >= healing.maxHeal * maxHP)
```

File: Plugins/Public/alley/PlayerRestrictions.cpp [hunk 7]
```diff
@@ -997,62 +1023,48 @@ bool ExecuteCommandString_Callback(CCmds* cmds, const wstring &wscCmd)
...
-						return;
-					}
-				}
-				//else do nothing, means it isn't a healing call.
-			}
-		}
+	float healedHP = (((healing.healingMultiplier / 100) + (curr / maxHP)) * maxHP) + healing.healingStatic;
+	float newHP = min(maxHP * healing.maxHeal, healedHP);
+
+	if (p1 == 1)
```
Comment (dsyalex) [inline Plugins/Public/alley/PlayerRestrictions.cpp:1058]: we're comparing some unknown variable to some magic constant? I may as well be looking at some FL assembly here

File: Plugins/Public/alley/PlayerRestrictions.cpp (hunk 2 above)
Comment (Aingar) [reply Plugins/Public/alley/PlayerRestrictions.cpp:1033]: ShortId of the component being hit. The magic number stands for shield, it's used in balancemagic as well.

File: Plugins/Public/alley/PlayerRestrictions.cpp (hunk 7 above)
Comment (Aingar) [reply Plugins/Public/alley/PlayerRestrictions.cpp:1058]: Will Adjust naming conventions and add appropriate comments.
1 stands for direct hull hit.

File: Plugins/Public/alley/PlayerRestrictions.cpp (hunk 3 above)
Comment (Aingar) [reply Plugins/Public/alley/PlayerRestrictions.cpp:1036]: iDmgToSpaceID is null on external equipment and collision group hits. Im 'fixing' the plugin by making it possible to fix the ship regardless of where on he hull the hit landed as long as it was not a shield hit.
Very impactful when dealing with new Battlecruisers on which a large chunk of their surface area are collision groups.

File: Plugins/Public/alley/PlayerRestrictions.cpp [hunk 8]
```diff
@@ -997,62 +1023,48 @@ bool ExecuteCommandString_Callback(CCmds* cmds, const wstring &wscCmd)
...
+	if (p1 == 1) 
+	{
+		damage = newHP;
+	}
+	else
+	{
+		iDmgTo = 0;
+		iDmgMunitionID = 0;
+		returncode = SKIPPLUGINS_NOFUNCTIONCALL;
+		dmg->add_damage_entry(1, newHP, static_cast<DamageEntry::SubObjFate>(0));
```
Comment (Aingar) [inline Plugins/Public/alley/PlayerRestrictions.cpp:1067]: Adding damage entry that will heal the main hull. 1 stands for hull internal id. NewHP is self-explanatory. Fate 0 means alive.

File: Plugins/Public/alley/PlayerRestrictions.cpp (hunk 1 above)
Comment (Aingar) [reply Plugins/Public/alley/PlayerRestrictions.cpp:263]: Yes, Haste took care of it.

File: Plugins/Public/alley/PlayerRestrictions.cpp (hunk 8 above)
Comment (dsyalex) [reply Plugins/Public/alley/PlayerRestrictions.cpp:1067]: This should be a comment in the code :)

**Pull request merged**